orderer/consensus/smartbft/chain_test.go

```diff
@@ -0,0 +1,709 @@
+	channelId string,
+	numberOfNodes int,
+	chainConfigFactory ChainConfigFactory,
+) (*TestNetwork, error) {
+	g := NewWithT(t)
+	workingDir := t.TempDir()
+
+	clusterState, err := NewTestClusterState(t, workingDir, numberOfNodes)
+	if err != nil {
+		return nil, err
+	}
+
+	genesisBlock := createMockGenesisBlock()
+	clusterState.AddBlockToAll(genesisBlock)
+
+	// Create chains
+	nodeIdToChain := map[uint64]*ChainUnderTest{}
+	nodeIdToActivity := map[uint64]bool{}
+
+	for nodeId, nodeState := range clusterState.NodeIdToState {
+		newChain, err := CreateNewChain(
+			t,
+			nodeState,
+			clusterState,
+			channelId,
+			nodeIdToChain,
+			chainConfigFactory,
+		)
+		if err != nil {
+			return nil, err
+		}
+		nodeIdToChain[nodeId] = newChain
+		nodeIdToActivity[nodeId] = true
+	}
+
+	return &TestNetwork{
+		t:                t,
+		g:                g,
+		clusterState:     clusterState,
+		nodeIdToChain:    nodeIdToChain,
+		ChannelId:        channelId,
+		nodeIdToActivity: nodeIdToActivity,
+	}, nil
+}
+
+func (n *TestNetwork) StartAllChains() {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	for nodeId := range n.nodeIdToChain {
+		n.startChainNoSync(nodeId)
+	}
+}
+
+func (n *TestNetwork) StartChain(nodeId uint64) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.startChainNoSync(nodeId)
+}
+
+func (n *TestNetwork) startChainNoSync(nodeId uint64) {
+	n.nodeIdToChain[nodeId].Start()
+	n.nodeIdToActivity[nodeId] = true
+}
+
+func (n *TestNetwork) StopAllChains() {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	for nodeId := range n.nodeIdToChain {
+		n.stopChainNoSync(nodeId)
+	}
+}
+
+func (n *TestNetwork) StopChain(nodeId uint64) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.stopChainNoSync(nodeId)
+}
+
+func (n *TestNetwork) stopChainNoSync(nodeId uint64) {
+	n.nodeIdToChain[nodeId].Halt()
+	n.nodeIdToActivity[nodeId] = false
+}
+
+func (n *TestNetwork) GetAgreedLeader() uint64 {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	var ids []uint64
+	for nodeId, c := range n.nodeIdToChain {
+		if !n.nodeIdToActivity[nodeId] {
+			continue
+		}
+		id := c.GetLeaderID()
+		ids = append(ids, id)
+	}
+	if allSameIds(ids) {
+		return ids[0]
+	}
+	return 0
+}
+
+func (n *TestNetwork) SendTxToAllChains(tx *cb.Envelope) error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	var errorsArr []error
+	for nodeId := range n.nodeIdToChain {
+		if !n.nodeIdToActivity[nodeId] {
+			errorsArr = append(errorsArr, nil)
+			continue
+		}
+		errorsArr = append(errorsArr, n.sendTxToChainNoSync(nodeId, tx))
+	}
+	return errors.Join(errorsArr...)
+}
+
+func (n *TestNetwork) sendTxToChainNoSync(nodeId uint64, tx *cb.Envelope) error {
+	if !n.nodeIdToActivity[nodeId] {
+		return fmt.Errorf("Node %d inactive", nodeId)
+	}
+	n.t.Logf("Sending TX to node %d", nodeId)
+	chain := n.nodeIdToChain[nodeId]
+	return chain.Order(tx, chain.NodeState.Sequence)
+}
+
+//////////////////////////////////////////////////////////////
+// ClusterState
+//////////////////////////////////////////////////////////////
```

Comment:
these kind of comments are not very useful. comment on the struct and explain whet is does